runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -97,12 +97,18 @@
             proxy.resetObjectUnsafe(object);
         }
 
-        // next, if the version is older than what we need
-        // sync.
-        if (object.getVersionUnsafe() < getSnapshotTimestamp()) {
-            proxy.syncObjectUnsafe(proxy.getUnderlyingObject(), getSnapshotTimestamp());
+        // Sync only if we never read the object in this transaction
+        // or if the version is older than what we need
+        OptimisticTransactionalContext oCtxt = (OptimisticTransactionalContext)
+                object.getModifyingContextUnsafe();
+        long snapshotOfThisStream =  streamSnapshots.computeIfAbsent(proxy,
+                k -> k.syncObjectUnsafe(k.getUnderlyingObject(), getSnapshotTimestamp()));
```

Comment:
This would happen only if the global tail address at snapshot time corresponds to the version number of the object, which means the latest global entry was from the object we are accessing.

We should get that case right as well. Will do the changes.